Add RFC for Stack metadata

kvedurmu opened the pull request:
[Readable](https://github.com/kvedurmu/rfcs/blob/stack-metadata/text/0000-stack-metadata.md)

hone commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+[motivation]: #motivation
+
+- Why should we do this?  
+Exposing additional metadata in stack images will make it easier for end users to debug stack related issues. Also, by exposing this metadata, CNB platforms such as kpack can leverage this information to improve their UX. 
```
What would `kpack` or a platform do with this data?

matthewmcnew replied:
`kpack` provides a `stack` resource that currently only reports the digest of the provided run/build image. It would be useful to report more detailed metadata status of that stack. That way users could better understand where the stack comes form. Something like:

```
Id:             io.buildpacks.stacks.alpine
Run Image:      alpine/run@sha256:d70bf0fe11d84277997c4a7da94b2867a90d6c0f55add4e19b7c565d5087206f
Build Image:    alpine/build@sha256:663a1781cb60a3feb43b9a2a2eae6681ca0124509b47719b5a6bf770c7638712
Mixins:         ca-certificates, libssl1.1
Url:            https://github.com/alpinelinux/cnb-alpine
ReleaseDate:    2020-05-12T05:17:02.390472
....
```

kvedurmu replied:
We could even expose some of this output in existing `pack` commands such as `rebase` if we wanted to.

ekcasey replied:
It would be nice to include this info in the `pack inspect-image` output as well

hone replied:
Can we include the kpack or pack examples in the RFC? :)

nebhale replied:
@hone Do the latest changes resolve this comment?

hone replied:
yep!

hone commented:
I assume this metadata is optional?

hone commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+
+- Any additional labels that we'd want to expose?
+
+# Spec. Changes (OPTIONAL)
```
Would we want to define this into a spec or extension spec?

kvedurmu replied:
It probably makes sense to define this in the core platform spec as optional fields here - https://github.com/buildpacks/spec/blob/master/platform.md#build-image

nebhale replied:
@hone I think this makes sense to go straight to core spec without a trial period in `extension`.  This is an augmentation to an already existent spec (labels on stack images) and we’re pretty clear on how it would be implemented and used.  I contrast this with the binding spec which is created from whole cloth with no understanding that it will work properly, be popular, and that all platforms will support it.

hone replied:
@nebhale I'm :+1: that. @kvedurmu do you mind adding updating this section with the changes you put in the platform spec?

hone replied:
Yeah, I'm :+1: of making it straight to the spec, I just wanted to make sure it was put in the RFC.

nebhale replied:
@kvedurmu I would written those into this RFC, but hey-ho a spec PR to start with will work too 😆

sclevine commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+`io.buildpacks.stack.release`: Release Number
+`io.buildpacks.stack.release_date`: Release date of image
+`io.buildpacks.stack.description`: Description  
+
```
Could we add a generic metadata field that is defined by the stack author?

E.g., `io.buildpacks.stack.metadata` could be defined by another out-of-spec RFC when the stack ID is io.buildpacks.stacks.bionic.

kvedurmu replied:
Yeah that makes sense. I think that we should keep the other labels I proposed since we already use this pattern for the `stack.id` and `stack.mixins` and add this generic metadata label. How does that sound?

nebhale replied:
Agreed on this.  I quite like the pattern of top-level keys being wholly reserved for the spec (i.e. `io.buildpacks.stack.*`) and having a key that is wholly reserved for users that the spec will never use (`io.buildpacks.stack.metadata`).  Top-level keys as they are described here look good.

hone replied:
> E.g., `io.buildpacks.stack.metadata` could be defined by another out-of-spec RFC when the stack ID is io.buildpacks.stacks.bionic.

@sclevine can you elaborate on what you hope to have in there?

ekcasey replied:
> I quite like the pattern of top-level keys being wholly reserved for the spec (i.e. io.buildpacks.stack.*) and having a key that is wholly reserved for users that the spec will never use (io.buildpacks.stack.metadata).

While I agree in theory, we currently use a lot of `io.buildpacks.*.metadata` for our own purposes, rather than reserving them for users. (e.g `io.buildpacks.{lifecycle,build, builder}.metadata`)

nebhale replied:
@ekcasey Well that's a kick in the shorts.  We can pick another key (`.user`) that represents the same thing, but alas is a naming inconsistency.  My priority stack is as follows:

1.  Designated area for all users data that we both agree to never spec inside of and that a user will never create outside of
2. Naming consistency

Even if that means we have some exceptions (e.g. `io.buildpacks{lifecycle,build,builder}.metadata`), I don't think we should introduce _new_ inconsistencies.

nebhale replied:
@kvedurmu I don't think this one has been addressed yet.

kvedurmu replied:
@nebhale oops just updated it.

kvedurmu commented:
> I assume this metadata is optional?
@hone yup, that's what I was thinking.

ekcasey commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+`io.buildpacks.stack.url`: URL for the stack (i.e repo link)  
+`io.buildpacks.stack.distro.name`: Name of OS Distribution  
+`io.buildpacks.stack.distro.version`: Version of OS Distribution  
+`io.buildpacks.stack.release`: Release Number
```
Maybe `version` here instead of `release`? It seems more consistent/generic

kvedurmu replied:
That seems reasonable. Updated

ekcasey commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+- Explaining the feature largely in terms of examples.  
+The following labels will be added to stack images:  
+`io.buildpacks.stack.maintainer`: Name of Stack Maintainer  
+`io.buildpacks.stack.url`: URL for the stack (i.e repo link)  
```
`homepage` would make this consistent with the buildpack metadata (assuming you mean git repo in the example here)

kvedurmu replied:
This makes sense to me. Updated.

nebhale commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+USER ${cnb_uid}:${cnb_gid}
+LABEL io.buildpacks.stack.id=${stack_id}
+LABEL io.buildpacks.stack.mixins="[\"ca-certificates\"............]"
+LABEL io.buildpacks.stack.maintainer="Paketo Buildpacks"
```
Thoughts on this being a Git-style string, e.g. `Paketo Buildpacks <stack@paketo.io>`?

kvedurmu replied:
I'm not sure I understand the purpose of a Git-style string for this. I feel like a combination of `io.buildpacks.stack.maintainer` plus `io.buildpacks.stack.homepage` tells the user who maintains the image and where to go for more information.

nebhale replied:
I'm not sure that `.homepage` implies contact information.  It points you to a place to learn about the stack, but not necessarily how to make contact with the maintainers.  I don't feel strongly on it though.

hone replied:
> Thoughts on this being a Git-style string, e.g. `Paketo Buildpacks <stack@paketo.io>`?

@nebhale I wonder if we should standardize this if we want to do that. We don't do it for [project.toml](https://github.com/buildpacks/spec/blob/master/extensions/project-descriptor.md#project).

nebhale replied:
Consistency first!  If we don't want to do it everywhere, let's not do it anywhere.

nebhale commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+
+- Define the target persona: stack author  
+- Explaining the feature largely in terms of examples.  
+The following labels will be added to stack images:  
```
Somewhere in the document, it should be called out that these are either optional or required.  I’m of the opinion that they should _all_ be optional, but don’t feel strongly about that.

hone replied:
optional but strongly encouraged?

nebhale replied:
My policy is MAY means WON'T and SHOULD means PROBABLY WON'T.  But  SHOULD is probably the right strength :)

nebhale commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,95 @@
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+None.
```
Given that this is a PR for a spec change, some sort of documentation needs to be added here, even it duplicates the discussion above.

nebhale commented:
@kvedurmu We'd like to get this moving again, can you please address the comments with new content?

kvedurmu commented:
Apologies @nebhale lost track of this. I believe I've addressed all the comments and [PRed](https://github.com/buildpacks/spec/pull/89) in a spec change for this. Let me know if I'm missing anything.

hone commented on text/0000-stack-metadata.md:
```diff
@@ -0,0 +1,138 @@
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This could make image rebasing slightly more complicated as labels like stack version will need to be updated.  
```
How would rebase work with this RFC?

matthewmcnew replied:
I believe the `io.buildpacks.stack.*` labels would need to be updated on the app image.

hone replied:
That sounds good to me. Can get how this works into the RFC itself?

hone commented:
I'm overall :+1: this RFC, but would like to see the rebase case addressed that was listed.

hone commented on text/0000-stack-metadata.md:
````diff
@@ -0,0 +1,138 @@
+LABEL io.buildpacks.stack.description="Paketo Buildpacks base stack build image"
+LABEL io.buildpacks.stack.metadata="Some optionally defined metadata"
+```
+
````
```suggestion
During rebase, the `rebaser` will copy the `io.buildpacks.stack.*` labels from the new run image to the app image.
```

nebhale commented:
Final Comment Period with merge disposition, closing on 22 July, 2020.

nebhale commented:
@kvedurmu Please do a rebase where you sign off on all the commits.
